contracts/protocol/modules/v1/DebtIssuanceModuleV3.sol

```diff
@@ -0,0 +1,241 @@
+            }
+            if(equityUnits[i] > tokenTransferBuffer) {
+                equityUnits[i] -= tokenTransferBuffer;
+            } else {
+                equityUnits[i] = 0;
+            }
+        }
+    }
+
+    /* ============ Internal Functions ============ */
+
+    function _getRequiredComponentIssuanceUnits(
+        ISetToken _setToken,
+        uint256 _quantity
+    )
+        internal
+        view
+        returns (address[] memory, uint256[] memory, uint256[] memory)
+    {
+        (
+            uint256 totalQuantity,,
+        ) = calculateTotalFees(_setToken, _quantity, true);
+
+        if(_setToken.totalSupply() == 0) {
+            return _calculateRequiredComponentIssuanceUnits(_setToken, totalQuantity, true);
+        } else {
+            (
+                address[] memory components,
+                uint256[] memory equityUnits,
+                uint256[] memory debtUnits
+            ) = _getTotalIssuanceUnitsFromBalances(_setToken);
+
+            uint256 componentsLength = components.length;
+            uint256[] memory totalEquityUnits = new uint256[](componentsLength);
+            uint256[] memory totalDebtUnits = new uint256[](componentsLength);
+            for (uint256 i = 0; i < components.length; i++) {
+                // Use preciseMulCeil to round up to ensure overcollateration of equity when small issue quantities are provided
+                // and use preciseMul to round debt calculations down to make sure we don't return too much debt to issuer
+                totalEquityUnits[i] = equityUnits[i].preciseMulCeil(totalQuantity);
+                totalDebtUnits[i] = debtUnits[i].preciseMul(totalQuantity);
+            }
+
+            return (components, totalEquityUnits, totalDebtUnits);
+        }
+    }
+
+
+    /**
+     * @dev Same as in v2 but adjusting the token transfers by the tokenTransferBuffer (adding when transferring in, subtracting when transferring out)
+     */
+    function _resolveEquityPositions(
+        ISetToken _setToken,
+        uint256 _quantity,
+        address _to,
+        bool _isIssue,
+        address[] memory _components,
+        uint256[] memory _componentEquityQuantities,
+        uint256 _initialSetSupply,
+        uint256 _finalSetSupply
+    )
+        internal
+        override
+    {
+        for (uint256 i = 0; i < _components.length; i++) {
+            address component = _components[i];
+            uint256 componentQuantity = _componentEquityQuantities[i];
+            if (componentQuantity > 0) {
+                if (_isIssue) {
+                    // Call SafeERC20#safeTransferFrom instead of ExplicitERC20#transferFrom
+                    SafeERC20.safeTransferFrom(
+                        IERC20(component),
+                        msg.sender,
+                        address(_setToken),
+                        // Add buffer to account for rounding errors
+                        componentQuantity + tokenTransferBuffer 
```

Comment:
Do we need this adjustment again? I thought it would already effect `componentQuantity` from this line https://github.com/IndexCoop/index-protocol/blob/a43caa279eacd77b7db7e2014cdf4eb2a1dde4a6/contracts/protocol/modules/v1/DebtIssuanceModuleV2.sol#L107